### DISCUSSION TITLE
Continous benchmark

### DISCUSSION THREAD

**Discussion by pfeatherstone** [davisking/dlib#2876] [Ideas]
Github actions supports continuous benchmark: https://github.com/marketplace/actions/continuous-benchmark
And it works with [Catch2](https://github.com/catchorg/Catch2)

I think this might be a cool thing to add. There have been a few PRs lately that improve performance. It would be nice if we could substantiate these claims with some graphs that show the evolution of benchmarks over commits and time.

This would potentially require rewriting some unit tests with Catch2, but I don't see that as a bad thing. It can be pulled in using cmake FetchContent, so it isn't a faffy dependency, and it would only be a dependency for unit tests, not the main library. The reporting is a bit nicer than dlib in my opinion anyway.

For example, i think tracking the performance of `dlib::matrix`, `dlib::fft`, and others are super important. Now i don't know if the results vary depending on the runner. Like, do Github runners occasionally run on different servers with different specs? That could skew results. I don't know.

**Comment by pfeatherstone** [comment]
Related discussion https://github.com/davisking/dlib/discussions/2767

**Comment by davisking** [comment]
Yeah I wager mightily the results vary a lot if run on GitHub. 

All the cloud services I've used have odd variability in runtimes if you measure them.  Although maybe it's not a big deal for this kind of thing.

Anyway. I'm more worried about dependency creep than anything else. So 🤷‍♂️ maybe

**Comment by davisking** [comment]
Anyway. I think the best thing is still to have some separate program that is specifically for benchmarking stuff. Like in the tools/ folder some benchmark program that when run outputs benchmarks. That would be fine and could be run wherever.

**Reply by pfeatherstone** [reply]
Yeah it would be nice if something somewhere ran that tool on every commit. And we could see graphs over time. Last thing you want is to accidentally merge something that kills performance.

**Reply by davisking** [reply]
Yeah, would be cool. I'm not super worried about it though.  But if you want to add a benchmark tool to the tools folder go for it.
